pkg/compactor/block_upload.go

```diff
@@ -559,3 +676,49 @@ func (c *MultitenantCompactor) loadValidation(ctx context.Context, userBkt objst
 
 	return v, nil
 }
+
+func (c *MultitenantCompactor) uploadValidationWithError(ctx context.Context, blockID ulid.ULID,
+	userBkt objstore.Bucket, errorStr string) error {
+	val := validationFile{
+		LastUpdate: time.Now().UnixMilli(),
+		Error:      errorStr,
+	}
+	dst := path.Join(blockID.String(), validationFilename)
+	if err := marshalAndUploadToBucket(ctx, userBkt, dst, val); err != nil {
+		return errors.Wrapf(err, "failed uploading %s to bucket", validationFilename)
+	}
+	return nil
+}
+
+func (c *MultitenantCompactor) uploadValidation(ctx context.Context, logger log.Logger, blockID ulid.ULID,
+	userBkt objstore.Bucket) error {
+	return c.uploadValidationWithError(ctx, blockID, userBkt, "")
+}
+
+func (c *MultitenantCompactor) periodicValidationUpdater(ctx context.Context, logger log.Logger, blockID ulid.ULID,
+	userBkt objstore.Bucket, ch chan string, wg *sync.WaitGroup, cancelFn func()) {
+	defer wg.Done()
+	for {
+		select {
+		case <-ch:
+			return
+		case <-time.After(1 * time.Minute):
+			if err := c.uploadValidation(ctx, logger, blockID, userBkt); err != nil {
+				level.Warn(logger).Log("msg", "error during periodic update of validation file")
+				cancelFn()
+				return
+			}
+		}
+	}
+}
+
+func marshalAndUploadToBucket(ctx context.Context, bkt objstore.Bucket, pth string, val interface{}) error {
```

Comment:
Could this function just be inlined? It's so simple, and it doesn't seem right to me to make a utility function just to share it with tests.

Reply:
Personally I think it serves better as a separate function, no need to clutter main logic with these details. (And can be reused)

Reply:
@pstibrany WDYT about my point of it being used from tests though, I think that makes the tests more difficult to reason about (i.e., bugs in a production function would affect test correctness).

Reply:
Reason for factoring out the code may also be to simplify main logic and document (by proper naming) what it does. I think this function does that.

While you have a point about reusing production code, I think this function is trivial enough to "see" its correctness, and hardly a core piece of tested functionality.

Reply:
I left this as-is for now.